### PR TITLE
Add style tag support

### PR DESCRIPTION
**Pull request by ptsavdar**

### CODE REVIEW

File: src/Rasterization/Renderers/SVGRenderer.php [hunk 1]
```diff
@@ -91,7 +91,7 @@ private static function prepareColor($color, SVGNode $context)
      */
     private static function getNodeOpacity(SVGNode $node)
     {
-        $opacity = $node->getStyle('opacity');
+        $opacity = $node->getComputedStyle('opacity');
```
Comment (JangoBrick) [inline src/Rasterization/Renderers/SVGRenderer.php]: This is wrong. Opacity is not inherited in CSS, but using `getComputedStyle()` would make it be inherited.

File: src/Nodes/Structures/SVGStyle.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * This file is part of the Claroline Connect package.
+ *
+ * (c) Claroline Consortium <consortium@claroline.net>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * Date: 2/8/17
+ */
```
Comment (JangoBrick) [inline src/Nodes/Structures/SVGStyle.php]: Where does this header come from?

File: src/Nodes/Structures/SVGDocumentFragment.php [hunk 3]
```diff
@@ -24,7 +25,6 @@ class SVGDocumentFragment extends SVGNodeContainer
     private $root;
     /** @var string[] $namespaces A map of custom namespaces to their URIs. */
     private $namespaces;
-
```
Comment (JangoBrick) [inline src/Nodes/Structures/SVGDocumentFragment.php]: Please leave empty lines as they are. If they need changing in your opinion, please submit an issue.

File: src/Nodes/SVGNodeContainer.php [hunk 4]
```diff
@@ -113,4 +143,58 @@ public function rasterize(SVGRasterizer $rasterizer)
...
+     *
+     * @param $pattern The pattern to search for, as a string
+     *
+     * @return array The matches array
+     */
+    private function pregGrepStyle($pattern)
+    {
+        $matches = preg_grep($pattern, array_keys($this->containerStyles));
+
+        return $matches;
```
Comment (JangoBrick) [inline src/Nodes/SVGNodeContainer.php]: No need for a temporary variable.

File: src/Rasterization/SVGRasterizer.php [hunk 5]
```diff
@@ -243,7 +243,7 @@ public function getHeight()
     public function getScaleX()
     {
         if (!empty($this->viewBox)) {
-            return $this->getDocumentWidth() / $this->viewBox[2];
+            return $this->width / $this->viewBox[2];
```
Comment (JangoBrick) [inline src/Rasterization/SVGRasterizer.php]: The document width and the rasterizer width are two very different things. Changing this doesn't seem like a good idea.

File: src/Utilities/SVGAttrParser.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * This file is part of the Claroline Connect package.
+ *
+ * (c) Claroline Consortium <consortium@claroline.net>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * Date: 2/10/17
+ */
+
```
Comment (JangoBrick) [inline src/Utilities/SVGAttrParser.php]: Again, what is this header?

File: src/Utilities/SVGAttrParser.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+namespace JangoBrick\SVG\Utilities;
```
Comment (JangoBrick) [inline src/Utilities/SVGAttrParser.php]: This class would've also made sense in `SVG\Reading`, don't you think?

File: src/Utilities/SVGStyleParser.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            }
+            $rules = self::parseStyles(trim($arr[2][$i]));
+            foreach ($selectors as $selector) {
+                $result[trim($selector)] = $rules;
+            }
+        }
+
+        return $result;
+    }
+}
```
Comment (JangoBrick) [inline src/Utilities/SVGStyleParser.php]: Please leave a trailing newline.

File: src/Rasterization/SVGRasterizer.php [hunk 9]
```diff
@@ -300,10 +297,7 @@ public function finish()
             return $this->outImage;
         }
 
-        $width  = $this->getDocumentWidth();
-        $height = $this->getDocumentHeight();
-
-        $this->outImage = self::clipImage($this->outImage, 0, 0, $width, $height);
+        $this->outImage = self::clipImage($this->outImage, 0, 0, $this->width, $this->height);
```
Comment (JangoBrick) [inline src/Rasterization/SVGRasterizer.php]: Same as above.

File: src/Nodes/Structures/SVGDocumentFragment.php [hunk 10]
```diff
@@ -106,7 +104,37 @@ public function getComputedStyle($name)
...
+            $svgViewBox,
+            $rasterizer->getWidth(),
+            $rasterizer->getHeight()
+        );
+        // Rasterize the svg and its children
+        parent::rasterize($svgRasterizer);
+        $img = $svgRasterizer->finish();
+        // Copy rasterized image to parent's image
+        imagecopy($rasterizer->getImage(), $img, 0, 0, 0, 0, $svgRasterizer->getWidth(), $svgRasterizer->getHeight());
+        imagedestroy($img);
```
Comment (JangoBrick) [inline src/Nodes/Structures/SVGDocumentFragment.php]: This is useful, thanks.

File: src/Nodes/SVGNodeContainer.php [hunk 11]
```diff
@@ -113,4 +145,58 @@ public function rasterize(SVGRasterizer $rasterizer)
...
+
+    /**
+     * Returns a style rules provided a given a node's id and class pattern.
+     *
+     * @param $pattern The node's id and class pattern for which we need to obtain
+     * its container style rules.
+     *
+     * @return array The style rules to be applied.
+     */
+    public function getContainerStyleForNodePattern($pattern)
```
Comment (JangoBrick) [inline src/Nodes/SVGNodeContainer.php]: Woah, quite the long name.

File: src/Rasterization/Renderers/SVGRenderer.php (hunk 1 above)
Comment (ptsavdar) [inline src/Rasterization/Renderers/SVGRenderer.php]: Ok I changed that.

File: src/Nodes/Structures/SVGStyle.php (hunk 2 above)
Comment (ptsavdar) [inline src/Nodes/Structures/SVGStyle.php]: Sorry about the headers. I've configured my IDE to add this automatically and forgot to delete it.

File: src/Nodes/Structures/SVGDocumentFragment.php (hunk 3 above)
Comment (ptsavdar) [inline src/Nodes/Structures/SVGDocumentFragment.php]: Again, I'm sorry about that. I've added trailing lines

File: src/Nodes/SVGNodeContainer.php (hunk 11 above)
Comment (ptsavdar) [inline src/Nodes/SVGNodeContainer.php]: I changed it to getContainerStyleByPattern

File: src/Nodes/SVGNodeContainer.php (hunk 4 above)
Comment (ptsavdar) [inline src/Nodes/SVGNodeContainer.php]: Ok fixed that too

File: src/Rasterization/SVGRasterizer.php (hunk 5 above)
Comment (ptsavdar) [inline src/Rasterization/SVGRasterizer.php]: Yes, you're right about the widths being 2 different things but I've tested with many images before my commit and found the rasterizer width to correspond better to the desired result than the document width. Below's an example with images:
Svg: 
image

Render with document width/height : 
image

Render with rasterizer width/height:
image

I've tested many other images, not just this one. You can try and test too if you want.

**Comment by ptsavdar** [comment]
I'd also like to thank you for taking my modifications into account and reviewing them. Here is my new PR following your advice/remarks #27 . Thanks again!

**Pull request closed without merging**